File: core/src/capi/plugins.zig
//! C-API Plugins Module - UI plugin discovery.
//!
//! Thin C-ABI glue over io.plugins.scanner. Scans ~/.talu/plugins/ and
//! returns an arena-backed list of discovered plugins.

const std = @import("std");
const capi_error = @import("error.zig");
const error_codes = @import("error_codes.zig");
const scanner = @import("../io/plugins/scanner.zig");

const allocator = std.heap.c_allocator;

// =============================================================================
// C-ABI Plugin Types
// =============================================================================

/// C-ABI plugin information record.
pub const CPluginInfo = extern struct {
    /// Plugin identifier (null-terminated).
    plugin_id: ?[*:0]const u8,
    /// Absolute path to the plugin directory (null-terminated).
    plugin_dir: ?[*:0]const u8,
    /// Relative entry path within plugin_dir (null-terminated).
    entry_path: ?[*:0]const u8,
    /// Raw JSON manifest string (null-terminated).
    manifest_json: ?[*:0]const u8,
    /// True if directory plugin, false if single-file.
    is_directory: bool,
    /// Reserved for future expansion.
    _reserved: [7]u8 = [_]u8{0} ** 7,
};

/// Plugin list container for C API.
pub const CPluginList = extern struct {
    /// Array of plugin info records.
    items: ?[*]CPluginInfo,
    /// Number of plugins in the array.
    count: usize,
    /// Internal: backing arena for string data.
    _arena: ?*anyopaque,
};

// =============================================================================
// List Builder
// =============================================================================

/// Build a CPluginList from scanned PluginInfo records.
fn buildPluginList(plugins: []const scanner.PluginInfo) !*CPluginList {
    const list = allocator.create(CPluginList) catch return error.OutOfMemory;
    errdefer allocator.destroy(list);

    const arena_ptr = allocator.create(std.heap.ArenaAllocator) catch return error.OutOfMemory;
    errdefer allocator.destroy(arena_ptr);
    arena_ptr.* = std.heap.ArenaAllocator.init(allocator);
    errdefer arena_ptr.deinit();

    const arena = arena_ptr.allocator();

    const items_buf = arena.alloc(CPluginInfo, plugins.len) catch return error.OutOfMemory;

    for (plugins, 0..) |plugin, i| {
        items_buf[i] = std.mem.zeroes(CPluginInfo);
        items_buf[i].plugin_id = (arena.dupeZ(u8, plugin.plugin_id) catch return error.OutOfMemory).ptr;
        items_buf[i].plugin_dir = (arena.dupeZ(u8, plugin.plugin_dir) catch return error.OutOfMemory).ptr;
        items_buf[i].entry_path = (arena.dupeZ(u8, plugin.entry_path) catch return error.OutOfMemory).ptr;
        items_buf[i].manifest_json = (arena.dupeZ(u8, plugin.manifest_json) catch return error.OutOfMemory).ptr;
        items_buf[i].is_directory = plugin.is_directory;
    }

    list.* = .{
        .items = if (plugins.len > 0) items_buf.ptr else null,
        .count = plugins.len,
        ._arena = @ptrCast(arena_ptr),
    };

    return list;
}

// =============================================================================
// Exported C API Functions
// =============================================================================

/// Scan a plugins directory and return discovered plugins.
///
/// Parameters:
///   - plugins_dir: Path to plugins directory (null-terminated).
///                  Pass null to use the default (~/.talu/plugins/).
///   - out_list: Output parameter to receive plugin list handle.
///
/// Returns: 0 on success, negative error code on failure.
/// On success, caller must free the handle with talu_plugins_list_free().
pub export fn talu_plugins_scan(
    plugins_dir: ?[*:0]const u8,
    out_list: ?*?*CPluginList,
) callconv(.c) i32 {
    capi_error.clearError();
    const out = out_list orelse {
        capi_error.setErrorWithCode(.invalid_argument, "out_list is null", .{});
        return @intFromEnum(error_codes.ErrorCode.invalid_argument);
    };
    out.* = null;

    const dir_slice = if (plugins_dir) |p| @as(?[]const u8, std.mem.span(p)) else null;
    const plugins = scanner.resolveAndScan(allocator, dir_slice) catch |err| {
        const code: error_codes.ErrorCode = switch (err) {
            scanner.ScanError.NoHomeDir => .invalid_argument,
            scanner.ScanError.OutOfMemory => .out_of_memory,
            else => .io_read_failed,
        };
        capi_error.setErrorWithCode(code, "Plugin scan failed: {s}", .{@errorName(err)});
        return @intFromEnum(code);
    };
    defer scanner.freePluginList(allocator, plugins);

    const list = buildPluginList(plugins) catch {
        capi_error.setErrorWithCode(.out_of_memory, "Failed to build plugin list", .{});
        return @intFromEnum(error_codes.ErrorCode.out_of_memory);
    };

    out.* = list;
    return 0;
}

/// Get the number of plugins in a list.
///
/// Parameters:
///   - list: Plugin list handle (may be null).
///
/// Returns: Number of plugins, or 0 if list is null.
pub export fn talu_plugins_list_count(list: ?*const CPluginList) callconv(.c) u32 {
    const l = list orelse return 0;
    return @intCast(l.count);
}

/// Get a plugin info record by index.
///
/// Parameters:
///   - list: Plugin list handle.
///   - index: Zero-based index into the list.
///
/// Returns: Pointer to plugin info, or null if out of bounds or list is null.
pub export fn talu_plugins_list_get(list: ?*const CPluginList, index: u32) callconv(.c) ?*const CPluginInfo {
    const l = list orelse return null;
    if (index >= l.count) return null;
    const items = l.items orelse return null;
    return &items[index];
}

/// Free a plugin list returned by talu_plugins_scan.
///
/// Parameters:
///   - list: Plugin list handle to free (may be null).
pub export fn talu_plugins_list_free(list: ?*CPluginList) callconv(.c) void {
    const l = list orelse return;

    if (l._arena) |arena_ptr| {
        const arena: *std.heap.ArenaAllocator = @ptrCast(@alignCast(arena_ptr));
        arena.deinit();
        allocator.destroy(arena);
    }

    allocator.destroy(l);
}

File: bindings/rust/talu/src/plugins.rs
//! Safe wrappers for talu UI plugin discovery.

use crate::error::error_from_last_or;
use crate::Result;
use std::ffi::{CStr, CString};
use std::path::{Path, PathBuf};

/// Discovered plugin information.
#[derive(Debug, Clone)]
pub struct PluginInfo {
    /// Plugin identifier (derived from dirname or filename without extension).
    pub plugin_id: String,
    /// Absolute path to the plugin directory.
    pub plugin_dir: String,
    /// Relative entry path within plugin_dir (e.g., "index.ts").
    pub entry_path: String,
    /// Raw JSON manifest string (talu.json content or inferred defaults).
    pub manifest_json: String,
    /// True if directory plugin, false if single-file.
    pub is_directory: bool,
}

impl PluginInfo {
    fn from_c(info: &talu_sys::CPluginInfo) -> Self {
        Self {
            plugin_id: c_str_to_string(info.plugin_id),
            plugin_dir: c_str_to_string(info.plugin_dir),
            entry_path: c_str_to_string(info.entry_path),
            manifest_json: c_str_to_string(info.manifest_json),
            is_directory: info.is_directory,
        }
    }
}

/// Return the default plugins directory (~/.talu/plugins/).
pub fn default_plugins_dir() -> PathBuf {
    if let Ok(dir) = std::env::var("TALU_PLUGINS_DIR") {
        return PathBuf::from(dir);
    }
    std::env::var_os("HOME")
        .map(PathBuf::from)
        .unwrap_or_else(|| PathBuf::from("."))
        .join(".talu")
        .join("plugins")
}

/// Scan a plugins directory and return discovered plugins.
///
/// If `plugins_dir` is `None`, uses the default (~/.talu/plugins/).
pub fn scan_plugins(plugins_dir: Option<&Path>) -> Result<Vec<PluginInfo>> {
    let c_dir = match plugins_dir {
        Some(dir) => {
            let s = dir.to_string_lossy();
            Some(CString::new(s.as_ref())?)
        }
        None => None,
    };

    let dir_ptr = c_dir.as_ref().map_or(std::ptr::null(), |s| s.as_ptr());

    let mut list: *mut talu_sys::CPluginList = std::ptr::null_mut();
    // SAFETY: dir_ptr is either null (use default) or a valid C string.
    // list is a valid out pointer.
    let rc = unsafe {
        talu_sys::talu_plugins_scan(dir_ptr, &mut list as *mut _)
    };
    if rc != 0 {
        return Err(error_from_last_or("Failed to scan plugins"));
    }

    let result = if list.is_null() {
        Vec::new()
    } else {
        // SAFETY: list is a valid pointer returned by talu_plugins_scan.
        let count = unsafe { talu_sys::talu_plugins_list_count(list) } as usize;
        let mut plugins = Vec::with_capacity(count);
        for i in 0..count {
            // SAFETY: index is within bounds.
            if let Some(info) = unsafe { talu_sys::talu_plugins_list_get(list, i as u32).as_ref() }
            {
                plugins.push(PluginInfo::from_c(info));
            }
        }
        plugins
    };

    // SAFETY: list was returned by talu_plugins_scan and must be freed.
    if !list.is_null() {
        unsafe { talu_sys::talu_plugins_list_free(list) };
    }

    Ok(result)
}

fn c_str_to_string(ptr: *const std::ffi::c_char) -> String {
    if ptr.is_null() {
        return String::new();
    }
    // SAFETY: ptr is a valid null-terminated C string from the C API.
    unsafe { CStr::from_ptr(ptr) }
        .to_string_lossy()
        .into_owned()
}
